### PR TITLE
ISPN-10354 Separate jboss-marshalling from core

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.jboss.org/browse/ISPN-10354

New approach that uses `JavaSerializationMarshaller` as the default user marshaller.

### CODE REVIEW

File: core/src/main/java/org/infinispan/marshall/core/LambdaMarshaller.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * @author Ryan Emerson
+ * @since 10.0
+ */
+class LambdaMarshaller {
+
+   static void write(ObjectOutput out, Object o) throws IOException {
+      try {
+         Method writeReplace = o.getClass().getDeclaredMethod("writeReplace");
+         writeReplace.setAccessible(true);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/core/LambdaMarshaller.java]: Do we need to worry about this with a security manager installed?

File: core/src/main/java/org/infinispan/distribution/ch/impl/ReplicatedConsistentHashFactory.java [hunk 2]
```diff
@@ -167,7 +167,7 @@ public Integer getId() {
 
       @Override
       public Set<Class<? extends ReplicatedConsistentHashFactory>> getTypeClasses() {
-         return Collections.<Class<? extends ReplicatedConsistentHashFactory>>singleton(ReplicatedConsistentHashFactory.class);
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/distribution/ch/impl/ReplicatedConsistentHashFactory.java]: Originally I had updated this class when converting to a protostream marshaller, then I reverted those changes but noticed that the type declaration wasn't required, so I updated that and changed to `Util.asSet` because that's what we have used in move of our Externalizers. I can change to just `Collections.singleton(..)` or  just not include this change if you prefer.

**Comment by gustavonalle** [comment]
Regarding configuration, to make it JSON friendly, I'd suggest the following:
* Having a pair ```WhiteListConfiguration```/```WhiteListConfigurationBuilder``` classes, centered around two attributes:
```
Attribute<List<Class<?>> classes = ...
Attribute<List<String> regexes = ...
```

* The XML would look like:
```xml
<white-list>
  <class>o.i.c.Class1</class>
  <class>o.i.c.Class2</class>
  <regex>regex1</regex>
  <regex>regex2</regex>
</white-list>
```

and the JSON will be rendered as:
```json
"white-list" : {
    "class":["o.i.c.Class1","o.i.c.Class2"],   
    "regex":["regex1","regex2"]
}
```

**Comment by wburns** [comment]
Btw appears to be related failures here https://ci.infinispan.org/job/Infinispan/job/PR-7175/7/#showFailuresLink

File: core/src/main/java/org/infinispan/distribution/ch/impl/ReplicatedConsistentHashFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/distribution/ch/impl/ReplicatedConsistentHashFactory.java]: I personally would prefer singleton, but looking at impl details of `Util.asSet` it actually uses singleton underneath if there was only a single entry. So I am fine either way.

File: core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java [hunk 3]
```diff
@@ -567,12 +626,13 @@ private void writeFlagsWithExternalizer(BytesObjectOutput out, Class<?> componen
...
    }
 
-   void writeUnknownClean(Marshaller marshaller, Object obj, ObjectOutput out) {
-      try {
-         writeUnknown(marshaller, obj, out);
-      } catch (IOException e) {
-         throw new CacheException(e);
-      }
+   private boolean isLambda(Class<?> clazz, Object obj) {
+      return clazz.isSynthetic() || obj instanceof SerializableFunction;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java]: Hrmm, what is the check for `SerializableFunction` here for? Is this the only serializable lambda we detect? Sadly there are a lot more :(

https://github.com/infinispan/infinispan/tree/master/core/src/main/java/org/infinispan/util/function

File: core/src/main/java/org/infinispan/marshall/exts/ThrowableExternalizer.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+import org.infinispan.remoting.transport.jgroups.SuspectException;
+import org.infinispan.statetransfer.AllOwnersLostException;
+import org.infinispan.statetransfer.OutdatedTopologyException;
+import org.infinispan.topology.CacheJoinException;
+import org.infinispan.transaction.WriteSkewException;
+import org.infinispan.transaction.xa.InvalidTransactionException;
+import org.infinispan.util.UserRaisedFunctionalException;
+import org.infinispan.util.concurrent.locks.DeadlockDetectedException;
+
+public class ThrowableExternalizer implements AdvancedExternalizer<Throwable> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/exts/ThrowableExternalizer.java]: I guess I never thought about it before, but this means we were using default Java serialization for Exceptions before? Technically this could have been done in a different PR though (although I wouldn't push that atm).

Either way great catch!

File: core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java [hunk 5]
```diff
@@ -57,21 +62,48 @@ public PersistenceMarshallerImpl() {
...
+         try {
+            log.jbossMarshallingDetected();
+            return clazz.getConstructor(GlobalComponentRegistry.class).newInstance(gcr);
+         } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            throw new CacheException("Unable to start PersistenceMarshaller with JBossUserMarshaller", e);
+         }
+      } catch (ClassNotFoundException e) {
+         ClassWhiteList whiteList = gcr.getCacheManager().getClassWhiteList();
+         UserMarshallerWhiteList.addInternalClassesToWhiteList(whiteList);
+         return new JavaSerializationMarshaller(whiteList);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java]: Maybe we should log an INFO message to notify them of the marshaller?

File: core/src/main/java/org/infinispan/marshall/persistence/impl/UserMarshallerWhiteList.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * A white list for Infinispan internal java classes which do not have a {@link org.infinispan.commons.marshall.Externalizer}
+ * instance defined.
+ *
+ * @author Ryan Emerson
+ * @since 10.0
+ */
+class UserMarshallerWhiteList {
+   static void addInternalClassesToWhiteList(ClassWhiteList list) {
+      // Infinispan Enums with no Externalizer
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/persistence/impl/UserMarshallerWhiteList.java]: I am surprised we don't have built in marshalling for Enums in our marshaller? It is just a simple name and ordinal to define any enum. We shouldn't have to add it to a whitelist in that case.

File: core/src/main/java/org/infinispan/marshall/persistence/impl/UserMarshallerWhiteList.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      list.addClasses(
+            BackupSender.BringSiteOnlineResponse.class,
+            BackupSender.TakeSiteOfflineResponse.class,
+            CacheMode.class,
+            XSiteAdminCommand.AdminOperation.class,
+            XSiteAdminCommand.Status.class
+      );
+      list.addClasses(Enum.class, Number.class, Object.class, Object[].class, SerializedLambda.class);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/persistence/impl/UserMarshallerWhiteList.java]: This would also allow removing Enum from this list. Object class isn't Serializable, why is this here?

File: core/src/main/java/org/infinispan/stats/impl/ClusterCacheStatsImpl.java [hunk 8]
```diff
@@ -552,4 +559,26 @@ private DistributedCacheStatsCallable(String cacheName) {
          return map;
       }
    }
+
+   public static class DistributedCacheStatsCallableExternalizer implements AdvancedExternalizer<DistributedCacheStatsCallable> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/stats/impl/ClusterCacheStatsImpl.java]: We can also remove `Serializable` from `DistributedCacheStatsCallable`

File: core/src/test/java/org/infinispan/filter/CompositeKeyValueFilterConverter.java [hunk 9]
```diff
@@ -15,6 +20,7 @@
  * @author wburns
  * @since 7.0
  */
+@SerializeWith(CompositeKeyValueFilterConverter.Externalizer.class)
 public class CompositeKeyValueFilterConverter<K, V, C> implements KeyValueFilterConverter<K, V, C>, Serializable, ExternalPojo {
```
Comment (wburns) [inline core/src/test/java/org/infinispan/filter/CompositeKeyValueFilterConverter.java]: This doesn't have to be `Serializable` anymore as well 👍

File: core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java [hunk 10]
```diff
@@ -56,7 +55,7 @@ protected void createCacheManagers() throws Throwable {
       configBuilder.clustering().hash().numSegments(4).stateTransfer().timeout(30000);
       globalBuilder = GlobalConfigurationBuilder.defaultClusteredBuilder();
       // Must explicitly add with id >= 2500 so that the user marshaller is able to store it
-      globalBuilder.serialization().addAdvancedExternalizer(JBossUserMarshaller.USER_EXT_ID_MIN, new JGroupsAddress.Externalizer());
+      globalBuilder.serialization().addAdvancedExternalizer(2500, new JGroupsAddress.Externalizer());
```
Comment (wburns) [inline core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java]: We should probably store the min somewhere again still, so we can reference if needed.

File: core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java [hunk 11]
```diff
@@ -41,6 +41,7 @@ protected void createCacheManagers() throws Throwable {
       builderUsed.clustering().hash().numOwners(1);
       builderUsed.memory().storageType(StorageType.BINARY);
       createClusteredCaches(3, builderUsed);
+      cacheManagers.forEach(cm -> cm.getClassWhiteList().addRegexps("java.util.Collections\\$SingletonMap"));
```
Comment (wburns) [inline core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java]: We should add `SingletonMap` to the `CollectionExternalizer`

File: core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java [hunk 12]
```diff
@@ -113,6 +121,7 @@ public void setMembersToUse(List<Address> membersToUse) {
       this.membersToUse = membersToUse;
    }
 
+   @SerializeWith(Externalizer.class)
```
Comment (wburns) [inline core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java]: Should we just add `ControlledConsistentHashFactory*` to the white list as you did with `BaseControlledConsistentHashFactory` above? Adds more code we have to worry about maintaining that buys us very little imo (since they are test classes).

Comment (wburns) [reply core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java]: Or I guess that could be that Address is not Serializable?

File: jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/io/GridFile.java [hunk 13]
```diff
@@ -1,4 +1,4 @@
-package org.infinispan.io;
+package org.infinispan.jboss.marshalling.io;
```
Comment (wburns) [inline jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/io/GridFile.java]: Why was all the GridFile stuff moved? As far as I saw, when I quickly searched, it wasn't using jboss marshalling.

File: jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/core/JBossUserMarshaller.java [hunk 14]
```diff
@@ -17,11 +19,12 @@
  * @author Ryan Emerson
  * @since 10.0
  */
+@SuppressWarnings("unused")
```
Comment (wburns) [inline jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/core/JBossUserMarshaller.java]: Seems extraneous

File: jboss-marshalling/src/test/java/org/infinispan/jboss/marshalling/dataconversion/DataConversionTest.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+      withCacheManager(new CacheManagerCallable(
+            createCacheManager(new ConfigurationBuilder())) {
+
+         GenericJBossMarshaller marshaller = new GenericJBossMarshaller();
+
+         private byte[] marshall(Object o) {
+            try {
+               return marshaller.objectToByteBuffer(o);
+            } catch (IOException | InterruptedException e) {
+               Assert.fail("Cannot marshall content");
```
Comment (wburns) [inline jboss-marshalling/src/test/java/org/infinispan/jboss/marshalling/dataconversion/DataConversionTest.java]: Maybe `throw new AssertionError(e)` instead?

File: commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java [hunk 16]
```diff
@@ -20,7 +18,7 @@
  * @since 4.0
  */
 @NotThreadSafe
-public final class ExposedByteArrayOutputStream extends ByteArrayOutputStream implements ByteOutput {
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java]: It's because `ByteOutput` is simply:

```java
public interface ByteOutput extends Closeable, Flushable {
   void write(int var1) throws IOException;

   void write(byte[] var1) throws IOException;

   void write(byte[] var1, int var2, int var3) throws IOException;
}
```

File: core/src/main/java/org/infinispan/configuration/global/WhiteListConfigurationBuilder.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import org.infinispan.commons.configuration.ClassWhiteList;
+import org.infinispan.commons.configuration.attributes.AttributeSet;
+
+/**
+ * Configures the {@link org.infinispan.manager.EmbeddedCacheManager} {@link org.infinispan.commons.configuration.ClassWhiteList}.
+ *
+ * @author Ryan Emerson
+ * @since 10.0
+ */
+public class WhiteListConfigurationBuilder implements Builder<WhiteListConfiguration> {
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/configuration/global/WhiteListConfigurationBuilder.java]: Previously the white list was only exposed via `EmbeddedCacheManager`, which meant that if anything need to be added to the list it had to happen after the manager has been created. Previously this wasn't important for embedded mode, as the white list was not used by the old user marshaller (`JbossUserMarshaller`), however it is a requirement of the `JavaSerializationMarshaller` so we should make it as flexible as possible for user configuration.

File: core/src/main/java/org/infinispan/marshall/core/InternalExternalizers.java [hunk 18]
```diff
@@ -260,6 +265,10 @@ static ClassToExternalizerMap load(GlobalComponentRegistry gcr, RemoteCommandsFa
       addInternalExternalizer(new SegmentPublisherResult.Externalizer(), exts);
       addInternalExternalizer(new PublisherReducers.PublisherReducersExternalizer(), exts);
       addInternalExternalizer(new CacheStreamIntermediateReducer.ReducerExternalizer(), exts);
+      addInternalExternalizer(new ClassExternalizer(gcr.getGlobalConfiguration().classLoader()), exts);
+      addInternalExternalizer(new ClusterCacheStatsImpl.DistributedCacheStatsCallableExternalizer(), exts);
+      addInternalExternalizer(ThrowableExternalizer.INSTANCE, exts);
+      addInternalExternalizer(new ImmutableListCopy.Externalizer(), exts);
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/marshall/core/InternalExternalizers.java]: Strictly speaking we don't in order for things to work with the JavaSerializationMarshaller, however we don't really want the marshalling of internal types to fall through to the user marshaller. So it makes sense that we handle this ourselves. Also, this is necessary for when plugging in a non-serialization based user marshaller i.e. a Protostream based marshaller.

File: core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java [hunk 19]
```diff
@@ -264,6 +286,7 @@ public static String requireSingleAttribute(final XMLStreamReader reader, final
         return requireAttributes(reader, true, attributeNames);
     }
 
+
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java]: Good spot, because actually none of the changes in this file are required ... I changed my approach based upon Gustavo's comments.

File: core/src/main/java/org/infinispan/marshall/exts/ThrowableExternalizer.java (hunk 4 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/marshall/exts/ThrowableExternalizer.java]: > I guess I never thought about it before, but this means we were using default Java serialization for Exceptions before?

Yep

Like the `ImmutableListCopy.Externalizer` this isn't strictly needed for the JavaSerializationMarshaller to work, but is for other non-serializable marshallers.

File: commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (wburns) [reply commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java]: Then why do we remove the interface?

File: core/src/main/java/org/infinispan/marshall/core/InternalExternalizers.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/marshall/core/InternalExternalizers.java]: I was referring to https://github.com/infinispan/infinispan/blob/master/tools/src/main/java/org/infinispan/tools/store/migrator/marshaller/infinispan8/ImmutableListCopyExternalizer.java But maybe it still needs to live for Infinispan8 migration.

File: core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java]: 👍

File: core/src/main/java/org/infinispan/configuration/global/WhiteListConfigurationBuilder.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/configuration/global/WhiteListConfigurationBuilder.java]: Thanks.

File: core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java (hunk 10 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java]: The `ControlledConsistentHashFactory.Default` now has it's own Externalizer, so no need fo rthis anyway. Any ideas on where to store the value?

Comment (ryanemerson) [reply core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java]: Unless we put it in the `AdvancedExternalizer` interface?

Comment (wburns) [reply core/src/test/java/org/infinispan/partitionhandling/NumOwnersNodeCrashInSequenceTest.java]: Seems like as fine of a place as any.

File: commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java]: Because it's from jboss-marshalling :smile:

[the remark this answers is not in the document]
Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java]: `org.jboss.marshalling.ByteOutput`

[the remark this answers is not in the document]
Comment (wburns) [reply commons/src/main/java/org/infinispan/commons/io/ExposedByteArrayOutputStream.java]: Ah 👍 For some reason I thought it was a standard java class, haha.

File: core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java (hunk 11 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java]: That's won't apply to this use-case, because the filter provided by the user (test) relies on Serializable and so both the filter and the map are handled by the user marshaller.

File: core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java (hunk 12 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java]: I have gone the other way and added Serializers for the `*ConsistentHash*` implementations. It's really internal code that shouldn't be handled by the user marshaller either way. I don't think maintenance effort is an issue, just some added boiler plate.

File: core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java (hunk 11 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java]: So I am confused, doesn't the user have some way of invoking our internal Marshaller classes? Or do they have to support Serialize every single class they may use?

File: jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/io/GridFile.java (hunk 13 above)
Comment (ryanemerson) [reply jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/io/GridFile.java]: Good point. I think I did it because `MarshalledValueByteStream` implemented `org.jboss.marshalling.ByteOutput`, but the move is overkill. Instead I think we can just remove the interface and users (if we have any) will have to call `{@link org.jboss.marshalling.Marshalling#createByteOutput(MarshalledValueByteStream)}` if they require it.

File: core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java (hunk 12 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/util/ControlledConsistentHashFactory.java]: yeah, I meant more it is just stuff to maintain and worry about in a test is all. I prefer when possible to have tests as specific as possible to not confuse what it is actually testing. But I understand if this has to be done.

File: jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/io/GridFile.java (hunk 13 above)
Comment (wburns) [reply jboss-marshalling/src/main/java/org/infinispan/jboss/marshalling/io/GridFile.java]: Sounds good, should reduce some noise in the PR.

File: core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java (hunk 11 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/stream/DistributedStreamIteratorWithStoreAsBinaryTest.java]: In general they shouldn't be using our internal classes. But it's possible todo so if they provide  externalizers for their classes.

File: core/src/main/java/org/infinispan/marshall/core/InternalExternalizers.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/marshall/core/InternalExternalizers.java]: O right, yeah we need that for the StoreMigrator.

File: core/src/main/java/org/infinispan/marshall/persistence/impl/UserMarshallerWhiteList.java (hunk 6 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/marshall/persistence/impl/UserMarshallerWhiteList.java]: We do now :smile:

File: core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java (hunk 3 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java]: I think the `isSynthetic` check is sufficient, as `instanceOf` those interfaces does not necessarily mean that it's a lambda, in which case we should not be using the LambdaMarshaller.

**Comment by ryanemerson** [comment]
@wburns I think I have addressed everything except docs. My plan was to write all of the docs around upgrading and marshalling nearer to `.Final` as I expect things to change a bit more in the mean time :smile:

File: core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java]: That is what I was thinking in regards to syntheic.

**Comment by ryanemerson** [comment]
`HotRodClientIT` failure should be fixed with the latest commit. Not rebasing yet as I want to make sure it doesn't break all of the server integration tests.

**Comment by ryanemerson** [comment]
run performance tests please

**Comment by wburns** [comment]
Integrated into master, thanks @ryanemerson !
